psydac/api/tests/test_api_2d_compatible_spaces.py

```diff
@@ -86,95 +91,126 @@ def run_system_1_2d_dir(f0, sol, ncells, degree):
     return l2_error
 
 #==============================================================================
-def run_system_2_2d_dir(f1, f2,u1, u2, ncells, degree):
-    # ... abstract model
-    domain = Square()
+def run_stokes_2d_dir(domain, f, ue, pe, *, homogeneous, ncells, degree, scipy=False):
 
+    # ... abstract model
     V1 = VectorFunctionSpace('V1', domain, kind='H1')
     V2 = ScalarFunctionSpace('V2', domain, kind='L2')
     X  = ProductSpace(V1, V2)
 
-    x,y = domain.coordinates
-
-    F = element_of(V1, name='F')
-
-    u,v = [element_of(V1, name=i) for i in ['u', 'v']]
-    p,q = [element_of(V2, name=i) for i in ['p', 'q']]
+    u, v = elements_of(V1, names='u, v')
+    p, q = elements_of(V2, names='p, q')
 
+    x, y  = domain.coordinates
     int_0 = lambda expr: integral(domain , expr)
 
-    
-    a  = BilinearForm(((u,p),(v,q)), int_0(inner(grad(u),grad(v)) + div(u)*q - p*div(v)) )
-    l  = LinearForm((v,q), int_0(f1*v[0]+f2*v[1]+q))
-    
-    bc = EssentialBC(u, 0, domain.boundary)
-    equation = find([u,p], forall=[v,q], lhs=a((u,p),(v,q)), rhs=l(v,q), bc=bc)
+    a  = BilinearForm(((u, p), (v, q)), int_0(inner(grad(u), grad(v)) - div(u)*q - p*div(v)) )
+    l  = LinearForm((v, q), int_0(dot(f, v)))
 
-    error = Matrix([F[0]-u1, F[1]-u2])
-    l2norm_F = Norm(error, domain, kind='l2')
+    if homogeneous:
+        bc = EssentialBC(u, 0, domain.boundary)
+    else:
+        bc = EssentialBC(u, ue, domain.boundary)
+
+    equation = find((u, p), forall=(v, q), lhs=a((u, p), (v, q)), rhs=l(v, q), bc=bc)
 
     # ... create the computational domain from a topological domain
     domain_h = discretize(domain, ncells=ncells)
+
     # ... discrete spaces
     V1h = discretize(V1, domain_h, degree=degree)
-    V2h = discretize(V2, domain_h, degree=degree)
+    V2h = discretize(V2, domain_h, degree=degree, basis='M')
+    Xh  = discretize(X , domain_h, degree=degree, basis='M')
```

Comment:
The basis argument 'M' only causes a renormalization of the basis functions for the 1D bases where the degree is reduced. 

The polynomial degree is reduced because we declared the space `V2` to be of kind `l2`.

There are lots of assumptions in the underlying construction, which should be made explicit and cleaned up...

Reply:
I understand. Indeed by looking at how V1 and V2 are defined and used, it seems that `kind` indicate the smoothness that is required on the functions. So yes, the assumptions in the construction should be clarified. 

This should probably be handled in a separate PR. I will update the issue #101 to include the construction of single spaces and the role of the `kind` parameter